tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,260 @@
+            application_name=application_name,
+            num_units=num_units,
+        )
+
+    await asyncio.gather(
+        deploy_wordpress(),
+        deploy_and_wait_for_mysql_pod(),
+        ops_test.run("playwright", "install", "chromium"),
+    )
+    await ops_test.model.applications[application_name].set_config(gen_upgrade_test_charm_config())
+    await ops_test.model.wait_for_idle(status=ops.model.ActiveStatus.name)  # type: ignore
+
+
+@pytest.mark.abort_on_fail
+async def test_create_example_blog(
+    ops_test,
+    default_admin_password,
+    unit_ip_list,
+    test_image,
+    screenshot_dir,
+    kube_core_client,
+    swift_config,
+    application_name,
+):
+    """
+    arrange: all charms and pods are deployed successfully.
+    act: enable some plugins and create posts and upload some images to the test WordPress blog.
+    assert: all operations finished without error.
+    """
+    namespace = ops_test.model_name
+
+    def get_wordpress_podspec_pod():
+        return (
+            kube_core_client.list_namespaced_pod(
+                namespace=namespace, label_selector=f"app.kubernetes.io/name={application_name}"
+            )
+            .items[0]
+            .metadata.name
+        )
+
+    wordpress_pod = get_wordpress_podspec_pod()
+
+    def kubernetes_exec(cmd):
+        logger.info("exec %s on %s", cmd, wordpress_pod)
+        resp = kubernetes.stream.stream(
+            kube_core_client.connect_get_namespaced_pod_exec,
+            name=wordpress_pod,
+            namespace=namespace,
+            container="wordpress",
+            command=cmd,
+            stderr=True,
+            stdin=False,
+            stdout=True,
+            tty=False,
+        )
+        logger.info("Response: %s", resp)
+
+    kubernetes_exec(
+        [
+            "curl",
+            "-sSL",
+            "https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar",
+            "-o",
+            "/usr/local/bin/wp",
+        ]
+    )
+    kubernetes_exec(["chmod", "+x", "/usr/local/bin/wp"])
+
+    def wp_cli_exec(wp_cli_cmd):
+        kubernetes_exec(wp_cli_cmd + ["--allow-root", "--path=/var/www/html"])
+
+    wp_cli_exec(["wp", "plugin", "activate", "openstack-objectstorage-k8s"])
+    wp_cli_exec(
+        ["wp", "option", "update", "object_storage", json.dumps(swift_config), "--format=json"]
+    )
+    unit_ip = unit_ip_list[0]
+    client = WordpressClient(
+        host=unit_ip,
+        username="admin",
+        password=default_admin_password,
+        is_admin=True,
+    )
+    media_id = client.upload_media(filename="wordpress.png", content=test_image)["id"]
+    post = client.create_post(title=POST_TITLE, content=POST_CONTENT, featured_media=media_id)
+    client.create_comment(post_id=post["id"], post_link=post["link"], content=POST_COMMENT)
+    for idx, unit_ip in enumerate(unit_ip_list):
+        await screenshot(f"http://{unit_ip}", screenshot_dir / f"wordpress-before-{idx}.png")
```

Comment:
This also looks like a fixture?

Reply:
Here should have an assertion that three screenshots from different units should be identical, but it hasn't been implemented yet.

Reply:
Changed to a fixture, thanks!